### ISSUE TITLE
[no-invalid-void-type] conditional types still get flagged as invalid voids

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#2549]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```js
module.exports = {
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "project": "tsconfig.json",
        "sourceType": "module"
    },
    "plugins": [
        "@typescript-eslint",
    ],
    "rules": {
       "@typescript-eslint/no-invalid-void-type": "error"
    }
};
```

```TS
export type SucceededDataResult<Data> = Data extends void
    ? {
        readonly succeeded: true;
    }
    : {
        readonly data: Data;
        readonly succeeded: false;
    };
```

**Expected Result**

No error message; this should be an acceptable use of `void`.

**Actual Result**

```
void is only valid as a return type or generic type variable
```

**Additional Info**

The rule should allow these kinds of conditional types, right? It seems reasonable that checking whether a type is `void` to, say, not include a member would be an ok act.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.1.0` |
| `@typescript-eslint/parser`        | `4.1.0` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.8.1` |
| `node`                             | `14.7.0` |

**Comment by bradzacher** [comment]
Why are you using void here instead of `undefined`?
It seems like the rule is reporting correctly?

(Also your type is probably wrong.. The failure case (`success: false`) includes the data)

**Comment by JoshuaKGoldberg** [comment]
> Why are you using void here instead of `undefined`?

It's intentional that this is meant to represent the _lack of_ return type, rather than a returned blank value (`undefined`) or a never-returning function (`never`).

> failure case (`success: false`) includes the data

Heh, nice spot, the actual type in code was originally a bit more complex before getting simplified for this issue. I think the code in the description doesn't do a good job of explaining the intent because of that typo, sorry! Here's a corrected version:

```ts
export type SucceededDataResult<Data> = Data extends void
    ? {
        readonly succeeded: true;
    }
    : {
        readonly data: Data;
        readonly succeeded: true;
    };
```

...and here's a slightly more complex system it's from:

```ts
export type DataResult<Data> = ErrorResult | SucceededDataResult<Data>;

export type ErrorResult = {
    error: Error;
    succeeded: false;
};
```

**Comment by tadhgmister** [comment]
> It seems reasonable that checking whether a type is void to, say, not include a member

> It's intentional that this is meant to represent the lack of return type,

How do you suggest the rule should identify this as correct? there is no indication that `Data` represents the return type of a function so checking if it is `void` makes sense. Also there is no viable way to see that the behaviour is to conditionally omit a field or not.

What change to the rule would let this case pass and not let other invalid uses also slip up? I feel like the best solution here would be to just have a `// eslint-disable-next-line` comment possibly with an explanation about what you are doing and leave the rule as is.

**Comment by JoshuaKGoldberg** [comment]
I'll just drop the subject then 😄 , looks like there's no strong feeling in support of doing something explicit for this admittedly rare case. Thanks all!
